### PR TITLE
Add capability to handle replication requests in vcr

### PR DESCRIPTION
**Pull request by ankagrawal**
This is part2 of the two part changes to enable two way replication. In a nutshell this change add capability in vcr to handle replication requests.

**NOTE: Atleast 48 of the 60 files changed, have changes only in import statement for ServerErrorCode, as the class was moved from ambry-commons to ambry-api as part of this change.**

 The main changes are:

- Refactor
1. Refactor StorageManager to create interface StoreManager, which is implemented by StorageManager.
2. Rename CloudBackupManager to VcrReplicationManager. Add StoreManager support for bookkeeping of cloudblobstores that a vcrserver is responsible for.
3. Move ErrorMapping and ServerMetrics classes to ambry-commons

- Create
1.  CloudStorageManager to manage cloudblobstores of a vcr server.
2. VcrRequests as a subclass of AmbryRequests to handle replication requests.
3. Fix tests.

### CODE REVIEW

**Comment by jsjtzyy** [comment]
@ankagrawal let's coordinate this PR with #1259 . I also made some changes in replication manager, storage manager and AmbryRequests.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrReplicationManager.java [hunk 1]
```diff
@@ -169,11 +171,9 @@ void addPartition(PartitionId partitionId) throws ReplicationException {
     }
     ReplicaId cloudReplica =
         new CloudReplica(cloudConfig, partitionId, virtualReplicatorCluster.getCurrentDataNodeId());
-    Store cloudStore = new CloudBlobStore(properties, partitionId, cloudDestination, clusterMap, vcrMetrics);
-    try {
-      cloudStore.start();
-    } catch (StoreException e) {
-      throw new ReplicationException("Can't start CloudStore " + cloudStore, e);
+    if (!storeManager.addBlobStore(cloudReplica)) {
```
Comment (jsjtzyy) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/VcrReplicationManager.java]: We discussed a little bit about similar logic of `addPartition` in `VcrReplicationManager` and `addReplica` in regular `ReplicationManager`. If we try to make it consistent and put single abstract method in base class, I feel like `addReplica` might be a more appropriate method name considering nature of `VCR` is actually a virtual replica. Helix controller assigns a replica of certain partition to this VCR node. In this `addPartition` method, the partitionId is used to create cloudReplica. We can move this out and pass in the created cloudReplica. Correct me if I am wrong.

**Comment by ankagrawal** [comment]
Sure. Lets sync up about it this afternoon after team meeting.

On Sep 16, 2019, at 10:20 AM, Yingyi Zhang <notifications@github.com<mailto:notifications@github.com>> wrote:


@ankagrawal<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fankagrawal&data=02%7C01%7Cankuagrawal%40linkedin.com%7Cd096ed3236e148058da008d73aca35b6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637042512471622869&sdata=DOgvrnb8Sf%2B7rp7M72DH1AxY5xqPsY%2FckexVN1HLgkM%3D&reserved=0> let's coordinate this PR with #1259<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Flinkedin%2Fambry%2Fpull%2F1259&data=02%7C01%7Cankuagrawal%40linkedin.com%7Cd096ed3236e148058da008d73aca35b6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637042512471632864&sdata=xYHCAoqPRBIWnb4DE9Ep4bsjYxARIjogf4U4PCi5Efc%3D&reserved=0> . I also made some changes in replication manager, storage manager and AmbryRequests.

File: ambry-api/src/main/java/com.github.ambry/server/StoreManager.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+
+public interface StoreManager {
+
+  /**
+   * Add a new BlobStore with given {@link ReplicaId}.
+   * @param replica the {@link ReplicaId} of the {@link Store} which would be added.
+   * @return {@code true} if adding store was successful. {@code false} if not.
+   */
+  public boolean addBlobStore(ReplicaId replica);
```
Comment (lightningrob) [inline ambry-api/src/main/java/com.github.ambry/server/StoreManager.java]: Don't need to specify public on interface methods.

File: ambry-api/src/main/java/com.github.ambry/server/StoreManager.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+   * Remove store from storage manager.
+   * @param id the {@link PartitionId} associated with store
+   * @return {@code true} if removal succeeds. {@code false} otherwise.
+   */
+  public boolean removeBlobStore(PartitionId id);
+
+  /**
+   * Start BlobStore with given {@link PartitionId} {@code id}.
+   * @param id the {@link PartitionId} of the {@link Store} which would be started.
+   */
```
Comment (lightningrob) [inline ambry-api/src/main/java/com.github.ambry/server/StoreManager.java]: doc return value

File: ambry-api/src/main/java/com.github.ambry/server/StoreManager.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  public Store getStore(PartitionId id);
+
+  /**
+   * Set BlobStore Stopped state with given {@link PartitionId} {@code id}.
+   * @param partitionIds a list {@link PartitionId} of the {@link Store} whose stopped state should be set.
+   * @param markStop whether to mark BlobStore as stopped ({@code true}) or started.
+   * @return a list of {@link PartitionId} whose stopped state fails to be updated.
+   */
+  public List<PartitionId> setBlobStoreStoppedState(List<PartitionId> partitionIds, boolean markStop);
```
Comment (lightningrob) [inline ambry-api/src/main/java/com.github.ambry/server/StoreManager.java]: The semantics of this method are weird (I realize you copied it).

File: ambry-api/src/main/java/com.github.ambry/server/StoreManager.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  /**
+   * Check if a certain partition is available locally.
+   * @param partition the {@link PartitionId} to check.
+   * @param localReplica the {@link ReplicaId} of the localreplica of the partition.
```
Comment (lightningrob) [inline ambry-api/src/main/java/com.github.ambry/server/StoreManager.java]: of the local replica

File: ambry-api/src/main/java/com.github.ambry/server/StoreManager.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+   * @return {@code true} if the partition is available. {@code false} if not.
+   */
+  public ServerErrorCode isPartitionAvailable(PartitionId partition, ReplicaId localReplica);//todo fix this
```
Comment (lightningrob) [inline ambry-api/src/main/java/com.github.ambry/server/StoreManager.java]: Return value doesn't match javadoc, and isXXX method typically returns boolean or throws exception.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import com.github.ambry.store.StoreException;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The storage manager that does book keeping for all the handled by this vcr node.
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: For all of the ... handled

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  @Override
+  public boolean removeBlobStore(PartitionId id) {
+    partitionTostore.remove(id);
+    return true;
+  }
+
+  @Override
+  public ServerErrorCode isPartitionAvailable(PartitionId partition, ReplicaId localReplica) {
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: See above.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrRequests.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Request implementation class for Vcr. All requests to the vcr server are
+ * handled by this class.
+ */
+public class VcrRequests extends AmbryRequests {
+
+  private Logger logger = LoggerFactory.getLogger(getClass());
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/VcrRequests.java]: Static logger (even if base class does it wrong).

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrServer.java [hunk 10]
```diff
@@ -70,6 +74,17 @@
...
+  private VcrRequests requests;
+  private RequestHandlerPool requestHandlerPool;
+
+  /**
+   * VcrServer constructor.
+   * @param properties the config properties to use.
+   * @param clusterAgentsFactory the {@link ClusterAgentsFactory} to use.
+   */
+  public VcrServer(VerifiableProperties properties, ClusterAgentsFactory clusterAgentsFactory) {
+    this(properties, clusterAgentsFactory, new LoggingNotificationSystem());
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/VcrServer.java]: Why is this needed?

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 11]
```diff
@@ -1123,21 +1122,18 @@ private ServerErrorCode validateRequest(PartitionId partition, RequestOrResponse
...
-          if (!storageManager.isDiskAvailable(localReplica.getDiskId())) {
-            metrics.diskUnavailableError.inc();
-            localReplica.markDiskDown();
-            return ServerErrorCode.Disk_Unavailable;
-          } else {
-            metrics.replicaUnavailableError.inc();
-            return ServerErrorCode.Replica_Unavailable;
-          }
-        } else {
+      ServerErrorCode errorCode = storeManager.isPartitionAvailable(partition, localReplica);
```
Comment (lightningrob) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: If you need the error code, call the method something like checkLocalPartitionStatus.

File: ambry-server/src/main/java/com.github.ambry.server/RequestHandlerPool.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      }
+      for (Thread thread : threads) {
+        thread.join();
+      }
+      logger.info("shut down completely");
+    } catch (Exception e) {
+      logger.error("error when shutting down request handler pool {}", e);
+    }
+  }
+}
```
Comment (lightningrob) [inline ambry-server/src/main/java/com.github.ambry.server/RequestHandlerPool.java]: newline

File: ambry-store/src/main/java/com.github.ambry.store/StorageManager.java [hunk 13]
```diff
@@ -157,16 +159,30 @@ public void start() throws InterruptedException {
...
     return diskManager != null ? diskManager.getStore(id) : null;
   }
 
+  @Override
+  public ServerErrorCode isPartitionAvailable(PartitionId partition, ReplicaId localReplica) {
+    if (getStore(partition) == null) {
+      if (localReplica != null) {
+        // check stores on the disk
+        if (!isDiskAvailable(localReplica.getDiskId())) {
+          localReplica.markDiskDown();
```
Comment (lightningrob) [inline ambry-store/src/main/java/com.github.ambry.store/StorageManager.java]: Looks like you're doing this twice.  Better to let caller do it since side effects are bad.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrServer.java [hunk 14]
```diff
@@ -155,9 +175,18 @@ public void startup() throws InstantiationException {
...
 
-      // TODO: for recovery, need AmbryRequests and RequestHandlerPool
+      //todo fix enableDataPrefetch
+      ServerMetrics serverMetrics = new ServerMetrics(registry, VcrRequests.class, VcrServer.class);
+      requests =
+          new VcrRequests(cloudStorageManager, networkServer.getRequestResponseChannel(), clusterMap, currentNode,
+              registry, serverMetrics, new FindTokenHelper(storeKeyFactory, replicationConfig), notificationSystem,
+              vcrReplicationManager, storeKeyFactory, true, storeKeyConverterFactory);
+
+      requestHandlerPool = new RequestHandlerPool(serverConfig.serverRequestHandlerNumOfThreads,
```
Comment (zzmao) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/VcrServer.java]: `requestHandlerPool` also needs shutdown.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrRequests.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+import java.io.IOException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Request implementation class for Vcr. All requests to the vcr server are
+ * handled by this class.
+ */
+public class VcrRequests extends AmbryRequests {
```
Comment (zzmao) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/VcrRequests.java]: Why extends `AmbryRequests` instead of implementing `RequestAPI`?
`VcrRequests` logic can be independent from  `AmbryRequests`, right?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrServer.java (hunk 14 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/VcrServer.java]: good catch, thanks!.. fixed

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrRequests.java (hunk 15 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/VcrRequests.java]: This is because except for a few helper utils that depend upon the idiosyncrasies of cloud, ambryrequests will be same as vcrrequests. (as discussed offline).

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudReplica.java [hunk 16]
```diff
@@ -72,7 +70,15 @@ public String getReplicaPath() {
...
   public List<ReplicaId> getPeerReplicaIds() {
-    return new ArrayList<>(partitionId.getReplicaIds());
+    List<ReplicaId> replicasOfPartition = new ArrayList<>();
+    for (ReplicaId replicaId : partitionId.getReplicaIds()) {
+      if (replicaId.getDataNodeId().getHostname().equals(dataNodeId.getHostname())
+          && replicaId.getDataNodeId().getPort() == dataNodeId.getPort()) {
+        continue;
+      }
+      replicasOfPartition.add(replicaId);
+    }
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudReplica.java]: This can be a stream - filter - collect expression.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java [hunk 17]
```diff
@@ -103,8 +106,7 @@ public boolean controlCompactionForBlobStore(PartitionId id, boolean enabled) {
 
   @Override
   public boolean removeBlobStore(PartitionId id) {
-    partitionTostore.remove(id);
-    return true;
+    return partitionTostore.remove(id) == null ? false : true;
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: Simpler: return partitionTostore.remove(id) != null;
Also, should be partitionToStore (camel case)

Comment (lightningrob) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: Never mind, already fixed.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+    return true;
+  }
+
+  @Override
+  public boolean controlCompactionForBlobStore(PartitionId id, boolean enabled) {
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public List<PartitionId> setBlobStoreStoppedState(List<PartitionId> partitionIds, boolean markStop) {
```
Comment (jsjtzyy) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: I don't know if the method is required by CloudStorageManager in the future. For now, I feel like it is more related to regular storage manager. How about removing this in interface and ClouldStorageManager? (It's easy to add this method if needed later)

**Comment by jsjtzyy** [comment]
One of the vcr tests failed. Please take a look.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 19]
```diff
@@ -1123,21 +1122,18 @@ private ServerErrorCode validateRequest(PartitionId partition, RequestOrResponse
...
+      switch (errorCode) {
+        case Disk_Unavailable:
+          metrics.diskUnavailableError.inc();
+          localReplica.markDiskDown();
```
[the remark this answers is not in the document]
Comment (ankagrawal) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: fixed.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java (hunk 11 above)
Comment (ankagrawal) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: fixed.

File: ambry-store/src/main/java/com.github.ambry.store/StorageManager.java (hunk 13 above)
Comment (ankagrawal) [reply ambry-store/src/main/java/com.github.ambry.store/StorageManager.java]: fixed.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+    throw new UnsupportedOperationException("Method not supported");
+  }
+
+  @Override
+  public boolean removeBlobStore(PartitionId id) {
```
[the remark this answers is not in the document]
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: remove method typically returns a boolean value for clients to safely check if the remove is successful, if they need to. An example is DiskManager::removeBlobStore. Interface signatures shouldn't depend upon implementation.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudReplica.java [hunk 21]
```diff
@@ -72,7 +70,13 @@ public String getReplicaPath() {
 
   @Override
   public List<ReplicaId> getPeerReplicaIds() {
-    return new ArrayList<>(partitionId.getReplicaIds());
+    List<ReplicaId> replicasOfPartition = partitionId.getReplicaIds()
+        .stream()
+        .filter(replica -> replica.getDataNodeId().getPort() != dataNodeId.getPort() || !replica.getDataNodeId()
+            .getHostname()
+            .equals(dataNodeId.getHostname()))
```
Comment (lightningrob) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudReplica.java]: You can simplify the filter condition to: replica.getDataNodeId().compareTo(dataNodeId) != 0

**Comment by jsjtzyy** [comment]
One unit test in `VcrRecoveryTest` failed. Let's fix it first

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+    if (cloudStore == null) {
+      cloudStore = new CloudBlobStore(properties, replica.getPartitionId(), cloudDestination, clusterMap, vcrMetrics);
+      partitionToStore.put(replica.getPartitionId(), cloudStore);
+    }
+    return startBlobStore(replica.getPartitionId());
+  }
+
+  @Override
+  public boolean shutdownBlobStore(PartitionId id) {
+    CloudBlobStore blobStore = partitionToStore.getOrDefault(id, null);
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: just `partitionToStore.get()` should return null if the element is not present.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java (hunk 18 above)
Comment (cgtz) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: There could potentially be some reasons for stopping replication from a vcr. I think it might be worth keeping in the interface.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+
+  @Override
+  public boolean removeBlobStore(PartitionId id) {
+    return partitionToStore.remove(id) != null;
+  }
+
+  @Override
+  public ServerErrorCode checkLocalPartitionStatus(PartitionId partition, ReplicaId localReplica) {
+    if (getStore(partition) == null) {
+      return ServerErrorCode.Replica_Unavailable;
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: Should this return Replica_Unavailable if a store is still in the map but has been shutdown?

In fact, the behavior of `getStore` in `CloudStorageManager` differs slightly from `DiskManager`. In `DiskManager` we do the following check:
```java
    BlobStore store = stores.get(id);
    return (running && store != null && store.isStarted()) ? store : null;
```

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrReplicationManager.java [hunk 24]
```diff
@@ -105,30 +108,10 @@ public void onPartitionAdded(PartitionId partitionId) {
...
-          if (cloudStore != null) {
-            cloudStore.shutdown();
-          } else {
-            logger.warn("Store not found for partition {}", partitionId);
-          }
-          logger.info("Partition {} removed from {}", partitionId, dataNodeId);
-          // We don't close cloudBlobStore because because replicate in ReplicaThread is using a copy of
-          // remoteReplicaInfo which needs CloudBlobStore.
+          removeReplica(partitionId);
+        } catch (ReplicationException e) {
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/VcrReplicationManager.java]: minor: these two `catch` clauses seem very similar. Could we just combine them?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrReplicationManager.java [hunk 25]
```diff
@@ -188,24 +168,27 @@ void addPartition(PartitionId partitionId) throws ReplicationException {
         FindTokenFactory findTokenFactory =
             tokenHelper.getFindTokenFactoryFromReplicaType(peerReplica.getReplicaType());
         RemoteReplicaInfo remoteReplicaInfo =
-            new RemoteReplicaInfo(peerReplica, cloudReplica, cloudStore, findTokenFactory.getNewFindToken(),
+            new RemoteReplicaInfo(peerReplica, cloudReplica, storeManager.getStore(partitionId),
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/VcrReplicationManager.java]: minor: save the result of `storeManager.getStore` to a variable and reuse it in this method.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/VcrReplicationManager.java [hunk 26]
```diff
@@ -242,6 +225,41 @@ void addPartition(PartitionId partitionId) throws ReplicationException {
...
+        persistor.write(partitionInfo.getLocalReplicaId().getMountPath(), false);
+      } catch (IOException | ReplicationException e) {
+        logger.error("Exception on token write when remove Partition {} from {}: ", partitionId, dataNodeId, e);
+        throw new ReplicationException("Exception on token write.");
+      }
+    }
+    Store cloudStore = partitionStoreMap.get(partitionId.toPathString());
+    if (cloudStore != null) {
+      try {
+        cloudStore.shutdown();
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/VcrReplicationManager.java]: Should the store be removed from `StoreManager` here or in some other place?
Also, should it be shut down using `StoreManager.shutdownBlobStore`?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java (hunk 23 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: StorageManager::checklocalPartitionStatus returns one of four error codes : Disk_Unavailable, Replica_Unavailable, Partition_Unknown and No_Error.
For CloudStoreManager, Disk_Unavailable doesn't make sense. Also, note that the only state that determines whether a vcr can serve a replica, is the state it gets from clustermanager. There is no disk for it to verify against. So I guess we either need to throw one of Replica_Unavailable or PArtition_Unknown in this case. Or we can also introduce a new state called Partition_Unassigned, as truly, this only means that the partition isnt assigned to this VCR, and we dont really know if Replica or Partition is unavailble. Let me know what you think of this.

Also fixed getStore.

File: build.gradle [hunk 27]
```diff
@@ -366,7 +366,8 @@ project(':ambry-cloud') {
             project(':ambry-store'),
             project(':ambry-commons'),
             project(':ambry-replication'),
-            project(':ambry-router')
+            project(':ambry-router'),
```
Comment (cgtz) [inline build.gradle]: What does ambry-cloud use the router dependency for?

Comment (ankagrawal) [reply build.gradle]: fixed.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java (hunk 23 above)
Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: fixed as discussed. we will return Partition_Unknown if the partition isnt assigned to vcr. We will return Replica_Unavailable if partition is stopped (blobstore is stopped).

File: build.gradle (hunk 27 above)
Comment (ankagrawal) [reply build.gradle]: only test required it. fixed.

**Comment by cgtz** [comment]
One of the integration tests seemed to get an NPE:
```
com.github.ambry.server.ServerSSLTest > endToEndCloudBackupTest[1] FAILED
    java.lang.NullPointerException
        at com.github.ambry.server.MockNotificationSystem.awaitBlobCreations(MockCluster.java:551)
        at com.github.ambry.server.ServerTestUtil.endToEndCloudBackupTest(ServerTestUtil.java:596)
        at com.github.ambry.server.ServerSSLTest.endToEndCloudBackupTest(ServerSSLTest.java:142)
```

File: ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+    this.cloudDestination = cloudDestination;
+    this.vcrMetrics = vcrMetrics;
+    this.clusterMap = clusterMap;
+  }
+
+  @Override
+  public boolean addBlobStore(ReplicaId replica) {
+    CloudBlobStore cloudBlobStore = partitionToStore.computeIfAbsent(replica.getPartitionId(),
+        store -> new CloudBlobStore(properties, replica.getPartitionId(), cloudDestination, clusterMap, vcrMetrics));
+    partitionToStore.put(replica.getPartitionId(), cloudBlobStore);
```
Comment (cgtz) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: I think that computeIfAbsent will already insert the entry into the map after computing so this call might not be needed

Comment (ankagrawal) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageManager.java]: fixed. also all tests passing.
